adapter/redisquota/redisquota.go

```diff
@@ -0,0 +1,315 @@
+	nanosPerTick = int64(time.Second / ticksPerSecond)
+)
+
+// Register records the builders exposed by this adapter.
+// TODO: need to be registered in inventory
+func Register(r adapter.Registrar) {
+	r.RegisterQuotasBuilder(newBuilder())
+}
+
+func newBuilder() builder {
+	return builder{adapter.NewDefaultBuilder(name, desc, conf)}
+}
+
+func (builder) NewQuotasAspect(env adapter.Env, c adapter.Config, d map[string]*adapter.QuotaDefinition) (adapter.QuotasAspect, error) {
+	return newAspect(env, c.(*config.Params), d)
+}
+
+// newAspect returns a new aspect.
+func newAspect(env adapter.Env, c *config.Params, definitions map[string]*adapter.QuotaDefinition) (adapter.QuotasAspect, error) {
+	dedupWindow, _ := ptypes.DurationFromProto(c.MinDeduplicationDuration)
+
+	return newAspectWithDedup(env, time.NewTicker(dedupWindow), c, definitions)
+}
+
+// newAspectWithDedup returns a new aspect.
+func newAspectWithDedup(env adapter.Env, ticker *time.Ticker, c *config.Params, definitions map[string]*adapter.QuotaDefinition) (adapter.QuotasAspect, error) {
+	connPool, _ := NewConnPool(c.SocketType, c.Url, c.PoolSize)
+	rq := &redisQuota{
+		definitions: definitions,
+		cells:       make(map[string]int64),
+		redisPool:   connPool,
+		recentDedup: make(map[string]dedupState),
+		oldDedup:    make(map[string]dedupState),
+		ticker:      ticker,
+		logger:      env.Logger(),
+	}
+	return rq, nil
+}
+
+func (rq *redisQuota) Alloc(args adapter.QuotaArgs) (adapter.QuotaResult, error) {
+	return rq.alloc(args, false)
+}
+
+func (rq *redisQuota) AllocBestEffort(args adapter.QuotaArgs) (adapter.QuotaResult, error) {
+	return rq.alloc(args, true)
+}
+
+func (rq *redisQuota) alloc(args adapter.QuotaArgs, bestEffort bool) (adapter.QuotaResult, error) {
+	amount, exp, err := rq.commonWrapper(args, func(d *adapter.QuotaDefinition, key string, currentTime time.Time, currentTick int64) (int64, time.Time,
+		time.Duration) {
+		result := args.QuotaAmount
+		conn, err := rq.redisPool.Get()
+		if err != nil {
+			return 0, time.Time{}, 0
+		}
+		defer rq.redisPool.Put(conn)
+
+		// increase the value of this key by the amount of result
+		conn.PipeAppend("INCRBY", key, result)
+		resp, err := conn.PipeResponse()
+
+		if err != nil {
+			return 0, time.Time{}, 0
+		}
+		ret := resp.Int()
+
+		if ret > d.MaxAmount {
+			if !bestEffort {
+				return 0, time.Time{}, 0
+			}
+			// grab as much as we can
+			result = d.MaxAmount - (ret - result)
+		}
+
+		return result, currentTime.Add(d.Expiration), d.Expiration
+	})
+
+	return adapter.QuotaResult{
+		Amount:     amount,
+		Expiration: exp,
+	}, err
+
+}
+
+func (rq *redisQuota) ReleaseBestEffort(args adapter.QuotaArgs) (int64, error) {
+	amount, _, err := rq.commonWrapper(args,
+		func(d *adapter.QuotaDefinition, key string, currentTime time.Time, currentTick int64) (int64, time.Time, time.Duration) {
+			result := args.QuotaAmount
+			conn, err := rq.redisPool.Get()
+			if err != nil {
+				return 0, time.Time{}, 0
```

Comment:
We shouldn't be dropping error returns, they must properly be propagated up. So quotaFunc should be expanded to expect an 'error' value and propagate it upwards accordingly.